### ISSUE TITLE
Exception: variable does not exist;

### ISSUE THREAD

**Issue by L619** [stan-dev/pystan-next#35]
I run this pystan 3.0 with the eight schools case, and find out an error:
File "/opt/python3/lib/python3.6/site-packages/stan/model.py", line 189, in build
raise RuntimeError(response_payload["error"]["message"])
RuntimeError: Error calling param_names: `Exception: variable does not exist; processing stage=data initialization; variable name=J; base type=int (in 'unknown file name' at line 3)

Could you please check it? Thank you.

The source code is :
import stan
program_code = """
data {
  int<lower=0> J; // number of schools
  real y[J]; // estimated treatment effects
  real<lower=0> sigma[J]; // s.e. of effect estimates
}
parameters {
  real mu;
  real<lower=0> tau;
  real eta[J];
}
transformed parameters {
  real theta[J];
  for (j in 1:J)
    theta[j] = mu + tau * eta[j];
}
model {
  target += normal_lpdf(eta | 0, 1);
  target += normal_lpdf(y | theta, sigma);
}
"""

data = {'J': 8,
        'y': [28,  8, -3,  7, -1,  1, 18, 12],
        'sigma': [15, 10, 16, 11,  9, 11, 10, 18]}

posterior = stan.build(program_code, data=data)
fit = posterior.sample(num_chains=4, num_samples=1000)

**Comment by riddell-stan** [comment]
Yes, I've been getting this myself in some virtualenvs but not others. I 
haven't been able to figure out what is happening.

The tests pass on travis with the code you provided (eight schools is 
one of the tests), so it's not something obvious.

Also, pystan 3 is not quite ready for testing. Try back in a week. 
(There's one or two changes in httpstan that need to be made.)

**Comment by L619** [comment]
Currently,the version of httpstan I installed is 0.6.1.  In your env,what is the version of httpstan?  Thank you.

**Comment by riddell-stan** [comment]
This is happening across versions for me.

**Comment by L619** [comment]
ERROR:aiohttp.server:Error handling request
Traceback (most recent call last):
File "/opt/python3/lib/python3.6/site-packages/aiohttp/web_protocol.py", line 390, in start
resp = await self._request_handler(request)
File "/opt/python3/lib/python3.6/site-packages/aiohttp/web_app.py", line 366, in _handle
resp = await handler(request)
File "/opt/python3/lib/python3.6/site-packages/httpstan/views.py", line 157, in handle_show_params
param_names_bytes = model_module.param_names(array_var_context_capsule)
File "model_22236aa30d.pyx", line 47, in model_22236aa30d.param_names
RuntimeError: Exception: variable does not exist; processing stage=data initialization; variable name=J; base type=int (in 'unknown file name' at line 3)

response <Response [500]>
Traceback (most recent call last):
File "test_pystan3.0.py", line 54, in <module>
test_func_small_case()
File "test_pystan3.0.py", line 46, in test_func_small_case
posterior = stan.build(program_code, data=data)
File "/opt/python3/lib/python3.6/site-packages/stan/model.py", line 185, in build
response_payload = response.json()
File "/opt/python3/lib/python3.6/site-packages/requests/models.py", line 897, in json
return complexjson.loads(self.text, **kwargs)
File "/opt/python3/lib/python3.6/json/__init__.py", line 354, in loads
return _default_decoder.decode(s)
File "/opt/python3/lib/python3.6/json/decoder.py", line 339, in decode
obj, end = self.raw_decode(s, idx=_w(s, 0).end())
File "/opt/python3/lib/python3.6/json/decoder.py", line 357, in raw_decode
raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)


This problem is still not solved. This is whole error info, could you please fix? Thank you.

**Comment by riddell-stan** [comment]
The provided code works for me with the current pystan-next branch (` d61542a`).

**Comment by L619** [comment]
After searching the branch, but I  don't find the version. Could you please give me link address of the version?  Thank you.

**Comment by riddell-stan** [comment]
Can you just try installing from `master`?

**Comment by L619** [comment]
The master version have same error.  I want to try your vesion(pystan-next branch  d61542a).  I don't I don't find it. Thank you.

**Comment by riddell-stan** [comment]
I'm getting this error with a fresh virtualenv. I'll keep trying to figure out what is going on.

**Comment by riddell-stan** [comment]
Try installing the wheel from pypi with:
```
pip install pystan==3.0.0.0a6
```

**Comment by riddell-stan** [comment]
I'm getting the error on one Linux system but not on another. Using a virtualenv in both cases. The error is occurring with Python 3.6.7 (Debian-based). It works fine on Python 3.6.5 (RHEL). ``pip freeze`` shows identical packages.

@ahartikainen can you try this out?

**Comment by ahartikainen** [comment]
I can try when I have time. 

This error is `httpstan` error not in `pystan`. In `pystan` is just fails to read empty data.

httpstan:

    File "/opt/python3/lib/python3.6/site-packages/httpstan/views.py", line 157, in handle_show_params
    param_names_bytes = model_module.param_names(array_var_context_capsule)
    File "model_22236aa30d.pyx", line 47, in model_22236aa30d.param_names
    RuntimeError: Exception: variable does not exist; processing stage=data initialization; variable name=J; 
    base type=int (in 'unknown file name' at line 3)

Notice

     param_names_bytes = model_module.param_names(array_var_context_capsule)

    RuntimeError: Exception: variable does not exist; 

    processing stage=data initialization; variable name=J; base type=int

Is it missing `;`? I hope there are no newline symbols or anything like that?

pystan:

    File "/opt/python3/lib/python3.6/site-packages/stan/model.py", line 185, in build
    response_payload = response.json()

To get better error message, it might mean that we need to first check if response can extract json?

In conclusion:

Maybe there is something wrong with the JSON library? Are there any external parts (other than Python).

**Comment by riddell-stan** [comment]
If I install from the tarball (from PyPI) instead of using the wheels, everything works on Python 3.6.7 (Debian-based). @L619 you might try that.

This still doesn't explain why Python 3.6.5 (RHEL) accepts the wheels and Python 3.6.7 (Debian-based) doesn't.

**Comment by riddell-stan** [comment]
I think this may be a problem with an incompatibility between the `manylinux` wheels and the freshly-compiled module. I could well be wrong.

Here's a possibly-relevant comment: https://github.com/pypa/manylinux/issues/118#issuecomment-308901809
>... If you have two extension modules that use C++ internally but that only communicate via python interfaces (i.e., they're not directly linking to each other or smuggling function pointers through pycapsules or something), then you should be OK with either static linking or vendoring libstdc++. (Watch out for cython's cimport functionality, which under the covers involves smuggling function pointers through pycapsules; otherwise I think you'll know if you're doing this.)

Since we *are* smuggling pointers via pycapsules between two extension modules that use C++ internally -- I'm assuming this might well be a problem.

Perhaps compiling with `-static-libstdc++` would help. Once I learn how to compile these manylinux wheels using docker locally, I'll give it a try.

In the short term, I'll make a new release of httpstan with no wheels (source only). I think this should fix things for @L619 .

**Comment by L619** [comment]
@riddell-stan, thank you, but for my stan model ,there is still other error, could you please help fix it? Thank you. 
The error is :
posterior = stan.build(program_code, data=dat)
File "/root/python3/lib/python3.6/site-packages/stan/model.py", line 186, in build
response = requests.post(f"http://{host}:{port}{path}", json=payload)
File "/root/python3/lib/python3.6/site-packages/requests/api.py", line 116, in post
return request('post', url, data=data, json=json, **kwargs)
File "/root/python3/lib/python3.6/site-packages/requests/api.py", line 60, in request
return session.request(method=method, url=url, **kwargs)
File "/root/python3/lib/python3.6/site-packages/requests/sessions.py", line 519, in request
prep = self.prepare_request(req)
File "/root/python3/lib/python3.6/site-packages/requests/sessions.py", line 462, in prepare_request
hooks=merge_hooks(request.hooks, self.hooks),
File "/root/python3/lib/python3.6/site-packages/requests/models.py", line 316, in prepare
self.prepare_body(data, files, json)
File "/root/python3/lib/python3.6/site-packages/requests/models.py", line 466, in prepare_body
body = complexjson.dumps(json)
File "/root/python3/lib/python3.6/json/__init__.py", line 231, in dumps
return _default_encoder.encode(obj)
File "/root/python3/lib/python3.6/json/encoder.py", line 199, in encode
chunks = self.iterencode(o, _one_shot=True)
File "/root/python3/lib/python3.6/json/encoder.py", line 257, in iterencode
return _iterencode(o, 0)
File "/root/python3/lib/python3.6/json/encoder.py", line 180, in default
o.__class__.__name__)
TypeError: Object of type 'Series' is not JSON serializable

The model is:

**Comment by L619** [comment]
The model is :
line_model_code = """
            data {
              int T;                                // Sample size
              int<lower=1> K;                       // Number of seasonal vectors
              real t[T];                            // Day
              real y[T];                            // Time-series
              int S;                                // Number of changepoints
              real A[T, S];                   // Split indicators
              real t_change[S];                 // Index of changepoints
              real X[T,K];                // season vectors
              vector[K] sigmas;              // scale on seasonality prior
              real<lower=0> tau;                  // scale on changepoints prior
            }

            parameters {
              real k;                            // Base growth rate
              real m;                            // offset
              real delta[S];                       // Rate adjustments
              real<lower=0> sigma_obs;               // Observation noise (incl. seasonal variation)
              real beta[K];                    // seasonal vector
            }

            transformed parameters {
              real gamma[S];                  // adjusted offsets, for piecewise continuity

              for (i in 1:S) {
                gamma[i] = -t_change[i] * delta[i];
              }
            }

            model {
              real Y[T];

              //priors
              k ~ normal(0, 5);
              m ~ normal(0, 5);
              delta ~ double_exponential(0, tau);
              sigma_obs ~ normal(0, 0.5);
              beta ~ normal(0, sigmas);

              // Likelihood
              for (i in 1:T) {
                Y[i] = (dot_product(A[i], delta) + k) * t[i] + (dot_product(A[i], gamma) + m) + dot_product(X[i], beta);
              }
              y ~ normal(Y, sigma_obs);
            }
            """

**Comment by riddell-stan** [comment]
Glad it fixed the problem. Thanks for bringing the problem to our attention.

You can't pass pandas Series as values in the `data` dictionary. We 
should add some checks here. I suspect many people will do what you do. 
But this is something for another issue.

**Comment by ahartikainen** [comment]
We could use `np.asarray` ?

**Comment by riddell-stan** [comment]
Let's continue this in #51. I also think we need a new issue for the C++ wheel issue.

**Comment by L619** [comment]
@riddell-stan ,But I tranform the dat to array, still have same error "TypeError: Object of type 'Series' is not JSON serializable". Could you please help fix it again? Thank you.

The modified code is(now seasonal_features is numpy.ndarray type):
        dat = {'T': history.shape[0],
            'K': seasonal_features.shape[1],
            'S': len(self.changepoints_t),
            'y': np.array(history['y_scaled'].values),
            't': np.array(history['t'].values),
            't_change': np.array(self.changepoints_t),
            'X': seasonal_features.values,
            'sigmas': np.array(prior_scales),
            'tau': self.changepoint_prior_scale,
            'trend_indicator': int(self.growth == 'logistic'),
            's_a': np.array(component_cols['additive_terms'].values),
            's_m': np.array(component_cols['multiplicative_terms'].values)
         }
